### PR TITLE
Support Gemma3 model

### PR DESCRIPTION
**Pull request by ForJadeForest**
1. Support Gemma3 model training 
2. Use label_key to get the ground truth answer in remote_reward serve. And support disable format reward
3. Add model parameter filtering in DeepspeedStrategy initialization.

### CODE REVIEW

File: openrlhf/models/lmm_kits/gemma3/data_processor.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        for inp in inputs:
+            batch.update({k: None for k, v in inp.items() if v is not None})
+        for k in batch.keys():
+            if k in ["input_ids", "attention_mask", "token_type_ids"]:
+                batch[k] = torch.stack([inp[k] for inp in inputs if k in inp], dim=0)
+            elif k in ["pixel_values"]:
+                # concat all patches of all images in a batch in the first dimension
+                batch[k] = torch.cat([inp[k] for inp in inputs if k in inp], dim=0)
+            else:
+                raise ValueError(f"Unknown key {k} for Qwen2VLDataProcessor")
```
Comment (TideDra) [inline openrlhf/models/lmm_kits/gemma3/data_processor.py:60]: Gemma3_VLDataProcessor

File: openrlhf/models/lmm_kits/gemma3/data_processor.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+                    batch_kwargs[i][k] = v
+        if "pixel_values" in keys:
+            pixel_values = batch["pixel_values"]
+            for i in range(batch_size):
+                token_type_ids_i = batch_kwargs[i]["token_type_ids"]
+                assert (token_type_ids_i == 1).sum() % 256 == 0
+                img_num = (token_type_ids_i == 1).sum().item() // 256
+                if img_num == 0:
+                    batch_kwargs[i]["pixel_values"] = None
+                    batch_kwargs[i]["image_grid_thw"] = None
```
Comment (TideDra) [inline openrlhf/models/lmm_kits/gemma3/data_processor.py:95]: gemma3有image_grid_thw这一项嘛

File: openrlhf/models/lmm_kits/gemma3/data_processor.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        messages = deepcopy(messages)
+        if isinstance(messages, list) and isinstance(messages[0], str):
+            formated_messages = [json.loads(m) for m in messages]
+        elif isinstance(messages, str):
+            formated_messages = [json.loads(messages)]
+        elif isinstance(messages, dict):
+            formated_messages = [[messages]]
+        else:
+            raise ValueError("Invalid messages format, must be a list of strings or a string or a dict")
+        formated_messages = self.warp_str_content_to_dict(formated_messages)
```
Comment (TideDra) [inline openrlhf/models/lmm_kits/gemma3/data_processor.py:135]: 上边直接用super()._format_messages吧

File: openrlhf/models/lmm_kits/gemma3/patch.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
+from ..base.patch import BasePatch
+import torch
+
+
+class Gemma3_Patch(BasePatch):
+    def _register_to_autoclass():
+        from transformers import AutoModel, Gemma3Config, Gemma3ForConditionalGeneration
+
+        AutoModel.register(Gemma3Config, Gemma3ForConditionalGeneration)
```
Comment (TideDra) [inline openrlhf/models/lmm_kits/gemma3/patch.py:9]: 现在是用AutoModelForImageTextToText加载模型了，如果gemma3在这个AutoMap里就不需要register了

File: openrlhf/models/lmm_kits/gemma3/patch.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                ):
+                    image_tokens_in_text = (special_image_mask).sum(dim=1).sum(dim=0)[0]
+                    raise ValueError(
+                        f"Number of images does not match number of special image tokens in the input text. "
+                        f"Got {image_tokens_in_text} image tokens in the text but {image_features.shape[0] * image_features.shape[1]} "
+                        "tokens from image embeddings."
+                    )
+                image_features = image_features.to(inputs_embeds.device, inputs_embeds.dtype)
+                inputs_embeds = inputs_embeds.masked_scatter(special_image_mask, image_features)
+            return inputs_embeds
```
Comment (TideDra) [inline openrlhf/models/lmm_kits/gemma3/patch.py:41]: pixel_values=None时要加fake pixel values

File: openrlhf/models/remote_rm/math_verifier.py [hunk 6]
```diff
@@ -95,60 +85,59 @@ def verify_math(content,sol):
...
     acc_rewards = [f.result() for f in acc_rewards_futures]
     rewards = [f + a for f, a in zip(format_rewards, acc_rewards)]
     # 返回包含 rewards 的响应
-    return jsonify({"rewards": rewards,"format_rewards":format_rewards,"acc_rewards":acc_rewards})
+    return jsonify({"rewards": rewards, "format_rewards": format_rewards, "acc_rewards": acc_rewards})
 
 
 if __name__ == "__main__":
     parser = ArgumentParser()
+    parser.add_argument("--dataset", type=str, default=None, help="Datasets to use (comma separated)", required=True)
```
Comment (TideDra) [inline openrlhf/models/remote_rm/math_verifier.py:124]: 用了label就不需要dataset了

**Review by TideDra — CHANGES_REQUESTED** [review]

**Review by TideDra — APPROVED** [review]
